components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sync/WorkManagerSyncManager.kt

```diff
@@ -511,13 +524,20 @@ internal class WorkManagerSyncWorker(
             }
         }
     }
+
+    companion object {
+        @VisibleForTesting
+        internal const val SYNC_STAGGER_BUFFER_MS = 5 * 1000L // 5 seconds.
+
+        @VisibleForTesting
+        internal val engineSyncTimestamp: MutableMap<String, Long> = mutableMapOf()
```

Comment:
I think the in-memory fix might work for now to avoid re-syncing while actively using the application. `companion object` is a thread-safe singleton so it'll live for the lifetime of the app. If the app is killed and we lose the in-memory mapping, then we've not used the application and synced a lot to worry about debouncing.

Reply:
If you do decide to persist the last synced values, please use the `SharedPreferencesCache` for serializing and deserializing.

Reply:
That was my thought as well.  Since closing the app and reopening will usually take some time, persisting the value is not needed yet.  When we add the API to allow user to customize custom debounce time, we should discuss if/how we persist them.

Reply:
> `companion object` is a thread-safe singleton lives for the lifetime of the app

I'm not willing to die on this hill, since this definitely seems the most straightforward way to get the desired behavior. IIUC, however, `static` (and therefore `companion`) objects and properties are not guaranteed to live in memory for the lifetime of an application. 

`companion` objects are generated as `static` members of their parent class, and will only remain in memory while the class definition is loaded. It is apparently unlikely for classes to ever be unloaded on a standard JVM, but I don't understand the ART runtime well enough to say the same. I would worry that Android's aggressive memory management is more likely to unload disused classes.

I would guess that the likelihood of that actually happening is incredibly low, if even possible. This seems especially true since I would guess that the regularity of syncs would keep these classes prioritized and loaded even _if_ any of my assumptions about ART are true

Reply:
I'm also curious about the `thread-safe singleton`. I think they are thread-safe during initialization, but I don't think the properties defined on them are. Since `doWork` is suspending on a `Dispatcher` with multiple threads, is it possible we could hit a deadlock without making this property `synchronized`?

Reply:
Thread safety is concerning.  I'll look into this.  Thanks

Reply:
Oh good, glad there was an easy fix for that.

Also, just for the sake of the conversation: my underlying issue here isn't about having persisted state of the timestamps. I just think it's a little code-smelly to be using a shared static property as a cache. Writing the state to disk felt like the easy solution to put the cache in a shared layer, which in this case would be disk storage. Something like an injected in-memory cache whose lifetime was controlled further up the object hierarchy could also work:

```
internal class WorkManagerSyncWorker(
    private val context: Context,
    private val params: WorkerParameters,
    private val syncTimestampCache: InMemoryCache,
) : CoroutineWorker(context, params) {

...
// the lifetime of this object is known to be longer, and can share the cache between periodic and one-time sync workers
internal class WorkManagerSyncDispatcher {
    private val timestampCache: InMemoryCache = mutableMapOf(<engines>, <timestamps>)
}
```